File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/Drive.java
package org.firstinspires.ftc.teamcode.drive;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

import org.firstinspires.ftc.teamcode.common.FtcGamePad;

public abstract class Drive {

    public enum TurnDirection {
        CLOCKWISE, COUNTER_CLOCKWISE
    }

    protected final LinearOpMode opMode;

    public Drive(LinearOpMode opMode) {
        this.opMode = opMode;
    }

    public abstract void drive(double v, double h, double r);
    public abstract void handleTeleop(FtcGamePad gamePad);
    public abstract void moveByInches(double power, float inches, double timeoutSeconds);
    public abstract void moveByInches(double power, float leftInches, float rightInches, double timoutSeconds);
    public abstract void turnDegrees(double power, TurnDirection turnDirection, int degrees, double timeoutSeconds);
    public abstract void encoderDrive(double power, int leftTicks, int rightTicks, double timeoutSeconds);
    public abstract void stop();

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/Drive.kt
package org.firstinspires.ftc.teamcode.drive

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.firstinspires.ftc.teamcode.common.FtcGamePad

abstract class Drive(opMode: LinearOpMode) {

    protected val opMode: LinearOpMode = opMode

    abstract fun drive(v: Double, h: Double, r: Double)
    abstract fun handleTeleop(gamePad: FtcGamePad)
    abstract fun moveByInches(power: Double, inches: Float, timeoutSeconds: Double)
    abstract fun encoderDrive(power: Double, leftTicks: Int, rightTicks: Int, timeoutSeconds: Double)
    abstract fun stop()

}